rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.hpp

```diff
@@ -400,6 +401,212 @@ bool TypeSupport<MembersType>::serializeROSmessage(
+      current_alignment += padding;
+      current_alignment += str_arr[index].size() + 1;
+    }
+  } else {
+    auto & data = *reinterpret_cast<std::vector<std::string> *>(field);
+    current_alignment += eprosima::fastcdr::Cdr::alignment(current_alignment, padding);
+    current_alignment += padding;
+    for (auto & it : data) {
+      current_alignment += eprosima::fastcdr::Cdr::alignment(current_alignment, padding);
+      current_alignment += padding;
+      current_alignment += it.size() + 1;
+    }
+  }
+  return current_alignment;
+}
+
+// C specialization
+template<typename T>
+size_t next_field_align(
+  const rosidl_typesupport_introspection_c__MessageMember * member,
+  void * field,
+  size_t current_alignment)
+{
+  const size_t padding = 4;
+  size_t item_size = sizeof(T);
+  if (!member->is_array_) {
+    current_alignment += eprosima::fastcdr::Cdr::alignment(current_alignment, item_size);
+    current_alignment += item_size;
+  } else if (member->array_size_ && !member->is_upper_bound_) {
+    current_alignment += eprosima::fastcdr::Cdr::alignment(current_alignment, item_size);
+    current_alignment += item_size * member->array_size_;
+  } else {
+    current_alignment += eprosima::fastcdr::Cdr::alignment(current_alignment, padding);
+    current_alignment += padding;
+
+    auto & data = *reinterpret_cast<typename GenericCArray<T>::type *>(field);
+    current_alignment += eprosima::fastcdr::Cdr::alignment(current_alignment, item_size);
+    current_alignment += item_size * data.size;
+  }
+  return current_alignment;
+}
+
+template<>
+inline
+size_t next_field_align<std::string>(
+  const rosidl_typesupport_introspection_c__MessageMember * member,
+  void * field,
+  size_t current_alignment)
+{
+  const size_t padding = 4;
+  using CStringHelper = StringHelper<rosidl_typesupport_introspection_c__MessageMembers>;
+  if (!member->is_array_) {
+    current_alignment = CStringHelper::next_field_align(field, current_alignment);
+  } else {
+    if (member->array_size_ && !member->is_upper_bound_) {
+      auto string_field = static_cast<rosidl_generator_c__String *>(field);
+      for (size_t i = 0; i < member->array_size_; ++i) {
+        current_alignment += eprosima::fastcdr::Cdr::alignment(current_alignment, padding);
+        current_alignment += padding;
+        current_alignment += strlen(string_field[i].data) + 1;
+      }
+    } else {
+      current_alignment += eprosima::fastcdr::Cdr::alignment(current_alignment, padding);
+      current_alignment += padding;
+      auto & string_array_field = *reinterpret_cast<rosidl_generator_c__String__Array *>(field);
+      for (size_t i = 0; i < string_array_field.size; ++i) {
+        current_alignment = CStringHelper::next_field_align(
+          &(string_array_field.data[i]), current_alignment);
+      }
+    }
+  }
+  return current_alignment;
+}
+
+template<typename MembersType>
+size_t TypeSupport<MembersType>::getEstimatedSerializedSize(
```

Comment:
Can you elaborate why you want to iterate over the complete message to get its size without serializing it?
Not criticizing, just wondering what's the use case.

Reply:
To avoid reallocations. If we know in advance the serialization size, internal buffer can be directly allocated to the correct size (with a call to reserve), and no reallocations will occur.